modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotTraverse.scala

```diff
@@ -1,50 +1,73 @@
 
-  def make[F[_]: MonadThrow](
+  def make[F[_]: Sync](
     loadInc: Hash => F[Option[Signed[GlobalIncrementalSnapshot]]],
     loadFull: Hash => F[Option[Signed[GlobalSnapshot]]],
     contextFns: SnapshotContextFunctions[F, GlobalSnapshotArtifact, GlobalSnapshotContext],
     rollbackHash: Hash
   ): GlobalSnapshotTraverse[F] =
     new GlobalSnapshotTraverse[F] {
+      val logger = Slf4jLogger.getLogger[F]
 
       def loadChain(): F[(GlobalSnapshotInfo, Signed[GlobalIncrementalSnapshot])] = {
-        def loadFullOrErr(h: Hash) = loadFull(h).flatMap(_.liftTo[F](new Throwable(s"Global snapshot not found, hash=${h.show}")))
-        def loadIncOrErr(h: Hash) = loadInc(h).flatMap(_.liftTo[F](new Throwable(s"Incremental snapshot not found, hash=${h.show}")))
-
-        def hashChain(h: Hash): F[NonEmptyChain[Hash]] =
-          loadInc(h).flatMap {
-            _.traverse { inc =>
-              hashChain(inc.lastSnapshotHash).map(_.append(h))
-            }.map(_.getOrElse(NonEmptyChain.one(h)))
+        def loadFullOrErr(h: Hash) =
+          loadFull(h).flatMap(_.liftTo[F](new Throwable(s"Global snapshot not found during rollback, hash=${h.show}")))
+        def loadIncOrErr(h: Hash) =
+          loadInc(h).flatMap(_.liftTo[F](new Throwable(s"Incremental snapshot not found during rollback, hash=${h.show}")))
+
+        def discoverHashesChain(rollbackHash: Hash): F[(Hash, NonEmptyChain[Hash])] =
+          (NonEmptyChain.one(rollbackHash), none[SnapshotOrdinal]).tailRecM {
+            case (hashes, lastOrdinal) =>
+              val lastHash = hashes.head
+
+              loadInc(lastHash)
+                .onError(_ => logger.error(s"Error during hash chain discovery at ${lastOrdinal.show} with hash ${lastHash.show}"))
+                .map {
+                  case Some(inc) => (hashes.prepend(inc.lastSnapshotHash), inc.ordinal.partialPrevious).asLeft
+                  case None      => (hashes, lastOrdinal).asRight
+                }
+          }.flatMap {
+            case (hashes, lastOrdinal) =>
+              val globalHashCandidate = hashes.head
+
+              NonEmptyChain.fromChain(hashes.tail) match {
+                case Some(incHashes) =>
+                  logger
+                    .info(s"Finished rollback chain discovery with global hash candidate $globalHashCandidate at $lastOrdinal")
+                    .as((globalHashCandidate, incHashes))
+                case None => RollbackSnapshotNotFound(rollbackHash).raiseError[F, (Hash, NonEmptyChain[Hash])]
+              }
           }
 
         for {
-          rollbackInc <- loadIncOrErr(rollbackHash)
-          (globalHash, incHashesNec) <- hashChain(rollbackInc.lastSnapshotHash).map { nec =>
-            (nec.head, NonEmptyChain.fromChainAppend(nec.tail, rollbackHash))
-          }
-          global <- loadFullOrErr(globalHash)
+          (globalHashCandidate, incHashesNec) <- discoverHashesChain(rollbackHash)
+          global <- loadFullOrErr(globalHashCandidate)
```

Comment:
Am I correct that the order of these operations has changed? Is that ok?

Reply:
I don't think the order changed. What is different here is that we don't need to try loading the rollback incremental hash first to get an error if it's not present. That will now happen in the `discoverHashesChain` function - a case where only one hash is there at the end in the NEC which effectively means that we didn't find rollback snapshot and that's why we don't have it's predecessor. Other than that I changed names a bit, because the globalHash is a candidate rather than a sure thing.